content/en/apps/guides/forms/rdtoolkit-integration.md

```diff
@@ -0,0 +1,219 @@
+---
+title: "Rapid Diagnostic Toolkit Integration"
+linkTitle: "RDToolkit Integration"
+weight:
+description: >
+  Integration for the Rapid Diagnostic Toolkit (RDToolkit)
+keywords: RDToolkit RDT
+
+---
+
+_Introduced in v3.12.0_
+
+CHT has an integration with [Rapid Diagnostic Toolkit (RDToolkit)](https://github.com/dimagi/rd-toolkit) which is an Android app developed by Dimagi that facilitates the use of Rapid Diagnostic Test (RDT) in global health care. 
+The integration is only available in the CHT Android app and consists of two Enketo widgets (rdtoolkit-provision and rdtoolkit-capture) that are used in XForms to provision RDT and capture results.
+
+## Create the XForms
+It is required to create one form for provisioning RDT and one form for capturing RDT results, see [the documentation](https://docs.communityhealthtoolkit.org/apps/tutorials/app-forms/) for more details about app forms.
+
+### Provisioning RDT form
+The following label is required in the `appearance` tag of `data` group:
```

Comment:
Originally I didn't have required groups, but then after some conversations about making specific selectors for the xml and the limited structure for making specific selectors without recurring to ids, though that must data were expected to be in "data" group.  
I'm going to look at the options to fulfil both concerns and comeback with my findings.

Reply:
Thanks, it's tricky, but worth keeping it as easy as possible for app developers and also maintaining compliance and styling with XForms and XLSForms.

Reply:
I think I've found a middle point:

We were searching by just field name prefixed by "rdtoolkit_" but there was a concern and moved it to specific selector:
Before:
 `$widget.find('input[name+="FIELD_NAME]"')`

Now it includes the form id and the group name `data`:
`$widget.find('input[name^="FORM_ID/data/FIELD_NAME]')`

[Recommended] We can relax the specific selector a bit, by looking for something that starts with the form id, omit anything in the middle (groups, or nested groups) and that ends with the field name, like a sandwich:
`$widget.find('input[name^="/FORM_ID"][name$="/FIELD_NAME]');`

Otherwise we might need to have our own group name: "rdtookit" 🙃

Note: we know what is the form id, it's variable provided by Enketo. I mean not hardcoded.

To confirm, is the sandwich solution okay? 🥪

Reply:
I am not certain of the best approach for the selector... but given that we do this in other widgets there should be a way to get this to work without hardcoding the field name. Perhaps this [section of the mrdt widget](https://github.com/medic/cht-core/blob/6e82151d120e57f2332d6cc1ea78da7a5bae69f9/webapp/src/js/enketo/widgets/mrdt.js#L49-L66) can serve as an example? Given @garethbowen's [work on the mrdt widget](https://github.com/medic/cht-core/commit/50c21bf39a17a5c242492c0359db451e46002061#diff-504326969274038d01a4fffb98edf8c4905c1be1f0e6545296a9fefe77e202c9) he may be a good resource for this.

Reply:
Looking at it, mrdt is using appearance tags as field name like: [or-appearance-mrdt-time-taken](https://github.com/medic/cht-core/blob/6e82151d120e57f2332d6cc1ea78da7a5bae69f9/webapp/src/js/enketo/widgets/mrdt.js#L61) to find and assign information into fields, but it's not really different as it has the same issue about not having specific selectors and more than one could match. 

I think the main difference with this integration is that we need to send more than 1 field and we need to receive more than 1 as well. Fields names won't be avoidable in this case as we need to make sure we're assigning the information in the right place.

I'm going to implement this solution that relaxes the selector a bit: `$widget.find('input[name^="/FORM_ID"][name$="/FIELD_NAME]');`

Reply:
By `[name$="/FIELD_NAME]'`, are you talking about the `or-appearance-*` name, or the XLSForm field name? If it is the `or-appearance-rdtoolkit-provision` then I think that's the more standard way to do it.

Reply:
`FIELD_NAME` is the element name in the XML tag and from the "name" column in the XLS, the standard in both.

<img width="500" alt="Screen Shot 2021-06-14 at 10 10 35 pm" src="https://user-images.githubusercontent.com/66472237/121915621-91ff1c80-cd5d-11eb-99b2-c293ac2a4fb1.png">

I think using appearance or any extra tag to identify fields is overhead, adds to complexity when building the forms.

Reply:
Using the appearance is the convention used in XLSForm for widgets like this, so I would follow that convention for the rdtoolkit group. The inner fields which are outputs can have set field names, as done in the external intent example.